File: 2024/day-02/day-02.fs
﻿module Day02

open FSharpx.Collections
open FSharpx.Text
open FSharpx
open Utils

let parse (input : string list) : int64 list list =
    input
    |> List.filter (not << Strings.isNullOrEmpty)
    |> List.map (Strings.split ' ' >> List.ofArray >> List.map int64)

let isSafe (level : int64 list) =
    let pairs = List.pairwise level
    let ascending = List.forall (uncurry (<)) pairs
    let descending = List.forall (uncurry (>)) pairs
    let maxDrift = List.fold (fun x (a, b) -> max x (abs (a - b))) 0L pairs
    (ascending || descending) && maxDrift <= 3L

let solvePart0 (input : string list) : int64 =
    let levels = parse input
    List.count isSafe levels

let solvePart1 (input : string list) : int64 =
    let levels = parse input

    let rec removeOnes =
        function
        | [] -> []
        | x :: xs -> xs :: (List.map (List.cons x) (removeOnes xs))

    List.count (List.exists isSafe << removeOnes) levels

type ThisDay() =
    interface IDay with
        member this.day () =
            Day.create 02 solvePart0 solvePart1
            <| seq {
                "input-ex0.txt", Some (2, 4)
                "input-real0.txt", Some (686, 717)
            }

File: 2015/day-02/day-02.c
#include "common/common.h"

static int64_t solvePart0([[maybe_unused]] Arena arena, FILE *f) {
    int64_t total = 0;
    int l, w, h;
    while (fscanf(f, "%dx%dx%d\n", &l, &w, &h) == 3) {
        total += 2 * l * w + 2 * w * h + 2 * h * l;
        total += MIN(l * w, MIN(w * h, h * l));
    }
    return total;
}

static int64_t solvePart1([[maybe_unused]] Arena arena, FILE *f) {
    int64_t total = 0;
    int l, w, h;
    while (fscanf(f, "%dx%dx%d\n", &l, &w, &h) == 3) {
        total += 2 * MIN(l + w, MIN(w + h, h + l));
        total += w * l * h;
    }
    return total;
}

static int dayMain() {
    int failed = 0;
    failed += checkInputInt("day-02/input-ex0.txt", solvePart0, 101, solvePart1, 48);
    failed += checkInputInt("day-02/input-real0.txt", solvePart0, 1606483, solvePart1, 3842356);
    return failed;
}

daySolver day02 = {2, dayMain, true};
